### PR TITLE
Flatness

### PR DESCRIPTION
**Pull request by DimpyRed**
The makes a lot of the area flat, but retains some deep hills and valleys.

### CODE REVIEW

**Comment by DimpyRed** [comment]
I want to rewrite the ways factors influence bock types now that flatness is a thing, but I want to make sure this is correct first.

File: client/src/worldgen.rs [hunk 1]
```diff
@@ -439,17 +445,38 @@ struct EnviroFactors {
...
+                        + rng.sample(&plus_or_minus_one)
+                        + ((parent.max_elevation < 0) as i64)
+                        - ((parent.max_elevation > 0) as i64)
+                }
+                31..=35 => {
+                    parent.max_elevation
+                        + (rng.sample(&plus_or_minus_one) + rng.sample(&plus_or_minus_one)) / 2
+                } //not sure about the rounding on this one
+                _ => parent.max_elevation,
+            },
```
Comment (Ralith) [inline client/src/worldgen.rs:479]: @MagmaMcFry, can you comment on whether this is sound?

Comment (DimpyRed) [reply client/src/worldgen.rs:479]: > Do we really need both "slopeiness" and "flatness"? Those sound like the same thing.

They are different. Slopiness functions as the derivitve of elevation change, while flatness functions as a suppressant for elevation change.

Comment (DimpyRed) [reply client/src/worldgen.rs:479]: In other words, slopeiness is a tool used to make terrain hilly, but flatness determines if it will actually be hilly.

**Comment by Ralith** [comment]
It sounds like they both indeed control the same thing. Could we do that with a single value instead?

File: client/src/worldgen.rs [hunk 2]
```diff
@@ -439,17 +445,26 @@ struct EnviroFactors {
...
+            .min(40);
         let slopeiness = parent.slopeiness + rng.sample(&plus_or_minus_one);
         Self {
             slopeiness,
+            flatness,
             max_elevation: parent.max_elevation
-                + ((3 - parent.slopeiness.rem_euclid(7)) + (3 - slopeiness.rem_euclid(7)))
-                + rng.sample(&plus_or_minus_one),
+                + ((((3 - parent.slopeiness.rem_euclid(7)) as f64)
+                    * (1.0 - (((parent.flatness as f64) - 20.0) / 10.0).tanh())
```
Comment (Ralith) [inline client/src/worldgen.rs:463]: `flatness - 20` has a range of -19 to 20 inclusive. That's asymmetrical, and I think therefore unsound; is it what you intended?

Comment (DimpyRed) [reply client/src/worldgen.rs:463]: It does not need to be symmetric. The 20.0 can be adjusted artistically. However, I might as well make it symmetric for neatness's sake.

Comment (Ralith) [reply client/src/worldgen.rs:463]: On review, the randomized sign at the end probably means any asymmetry here gets worked out in the end, but I still suspect it wasn't intended.

Comment (DimpyRed) [reply client/src/worldgen.rs:463]: Besides, 1 - tanh(x) is always more than zero.

**Pull request merged**